### DISCUSSION TITLE
The provided `href` value is missing query values to be interpolated properly

### DISCUSSION THREAD

**Discussion by gigamesh** [vercel/next.js#33243] [Help]
I'm getting this error on fresh page loads (_not_ from the Link component or router methods):

`Error: The provided `href` (/artist/release/[releaseId]?) value is missing query values (releaseId) to be interpolated properly.`

The URL is: `http://localhost:3000/artist/release/6480b71f-e8d4-4e9c-b672-c47690b69662`
The file path is `/artist/release/[releaseId].tsx`

Everything seems correct to me. Any idea what the source of the problem might be?

**Comment by johnbabu021** [comment]
https://nextjs.org/docs/messages/href-interpolation-failed
it may help

**Reply by gigamesh** [reply]
That page only describes when "you are utilizing the next/link component, Router#push, or Router#replace with href interpolation to build dynamic routes but did not provide all of the needed dynamic route params to properly interpolate the href."

In my case, I'm getting it from a fresh page load.

**Comment by ack210** [comment]
I am having the same issue too.  When I look at the call stack in dev tools, the error seems to be resulting from [this call to `router.replace` in `next-dev.js`](https://github.com/vercel/next.js/blob/d7a96d22be60bd84297befe7bff9ecc8862eaebf/packages/next/client/next-dev.js#L84), which is called from [this method](https://github.com/vercel/next.js/blob/d7a96d22be60bd84297befe7bff9ecc8862eaebf/packages/next/client/dev/error-overlay/websocket.ts#L72) as a result of an `serverOnlyChanges` event.

The `router.pathname` value the method is called with is, e.g. `/blog/[articleId]/[commentId]` as opposed to, e.g. `/blog/123/456`.

I should add that the page renders correctly underneath the error overlay (including data from api calls that are dependent on the slug values), and I don't get the error every time... I've struggled to find a way to reproduce it reliably so far.

**Reply by tzvc** [reply]
Running into the exact same issue

**Reply by amitchhatbar11** [reply]
> Running into the exact same issue

Found any solution? I am facing the same issue on hot reload. @theochampion

**Reply by avinashbarfa-buku** [reply]
Any resolution ? Facing Similar on refresh page

**Reply by amitchhatbar11** [reply]
@avinashbarfa-buku 
Which version of next you're using?

I resolved it by updating it to `12.1.0`.

**Reply by avinashbarfa-buku** [reply]
Using same version.
`"next": "12.1.0"`

**Reply by gustavos-br** [reply]
Having the same issue here using next@12.1.4

**Reply by wilbsy** [reply]
Having same issue as well (next@12.1.4).  Any solution?

**Reply by zachlankton** [reply]
I was also having this same problem today.  The answer was not obvious either, but the linked article does in fact direct us to the cause of this problem.  After some digging and commenting out various lines, I finally found what section of my code was generating the error. 

For me it was actually in an imported module that I was using in the page.

In my `/parts-tracking/[scenario]` route I was using a filter hook that pushes its search parameters to the url for bookmarking purposes using `router.push`

---

The old code:

```javascript
    router.push("?search=" + encodeURIComponent(val), undefined, {
      shallow: true,
    });
```

The new code that fixed the problem for me:

```javascript
   router.push(
      {
        query: {
          ...router.query,
          search: val,
        },
      },
      undefined,
      {
        shallow: true,
      }
    );
```

**Reply by wilbsy** [reply]
I found the issue in my code.  It was caused by the `router.push(router.pathname)` in the fix for framer-motion page transition issue [detailed here #17464 @ issuecomment-949620547](https://github.com/vercel/next.js/issues/17464#issuecomment-949620547).

**Reply by Jesuscc9** [reply]
but the problem is when you refresh on that route, not when you make a router push.

**Reply by wilbsy** [reply]
You can use `router.asPath` (as opposed to pathname) to get the true URL

**Reply by tymzd** [reply]
> You can use `router.asPath` (as opposed to pathname) to get the true URL

This solved it for me, thank you!

**Reply by oversio** [reply]
I still getting the same error when I refresh on the route, someone found a solution for that? 🙏

**Reply by TechnoRahmon** [reply]
I am getting the same issue when refreshing the dynamic route, any solution yet?

**Reply by mitrakmt** [reply]
No one with a solution yet? I'm starting to hate NextJS

**Reply by marcelo-earth** [reply]
Any update on this?

**Reply by matiassanchez12** [reply]
any solution?

**Reply by AnshulShirbhate** [reply]
I have a solution for this problem. So It was basically arising when I was not giving a value in the href attribute of the Link tag of Next.JS. As soon as removed that Link the error disappeared. Replace the Link tag where not needed or fill the href attribute with some value. 100% Working✅✅✅✅

**Comment by TamLa21297** [comment]
try add this function, it work for me:

export async function getServerSideProps(context) {
    return {
        props: {},
    };
}

**Reply by mbappai** [reply]
You should probably mention the reason why this worked for?
[Check out this next doc page](https://nextjs.org/docs/advanced-features/automatic-static-optimization) for more info about the above solution

**Reply by TechnoRahmon** [reply]
@TamLa21297  I am using `getServerSIdeProps ` and it does nothing to the issue

**Reply by vehere-ccuMaksym** [reply]
It converst the page into SSR one, but does not solves the issue for CSR pages

**Comment by serhanguney** [comment]
This one helped me but not sure if this is your case:
If you're using `useRouter` anywhere on that page watch out for the following:

👉 In dynamic routes the pathname value in `useRouter` returns the file name instead of the pathname. I did some checking with `query.slug` in such cases and it resolved my issue:
```js
const { pathname , query } = useRouter() // pathname is /artist/release/[releaseId] in your case

// my solution:
// relativePath will be artist/release/6480b71f-e8d4-4e9c-b672-c47690b69662

const relativePath = typeof query.slug === 'string' ? pathname.replace('[slug]', query.slug) : pathname
```

The issue was in my navbar where I had internalization control with a dropdown that displays either english or turkish version of the website. Based on the choice the locale value changes from `en` to `tur` and the pathname follows `/tur/[pathname]`...

The issue, of course, was that I ended up with the following anchor tag
```js
<a href="/tur/artist/release/[releaseId]">tur</a>
```

which should have been
```js
<a href="/tur/artist/release/6480b71f-e8d4-4e9c-b672-c47690b69662">tur</a>
```

**Comment by AayushKarki714** [comment]
for me this came up, when I forget to pass all the params needed by the page, 

router.replace({
         query: {
          ...router.query,
          callbackURL: `/event/${slug}/event-registration`,
           },
           });

**Comment by ruzolut** [comment]
If you are passing `pathname` to a router push or replace method, I suggest you pass `asPath` from the router object instead of `pathname` when you have to navigate the user to a path that has query/params.

In my case, I had this error when I was trying to change the locale using `router.push(...)`, I passed `pathname` as the first parameter to that method, I fixed it by doing this:
```ts
  conat router = useRouter()
  const { asPath, pathname, locale, push } = router

  const handleChangeLocale = async (locale: string) => {
    push(asPath, undefined, { locale }) // this one works
    // passing pathname instead doesn't work and returns your error
    // because the placeholders weren't replaced with their values yet.
  }
```

**Comment by MarkForLoop** [comment]
I am having the exact same error message, when I refresh the page on next 14.2.3

I was able to fix it by checking if the route was ready with `router.isReady`.

I don't know if this is an ok solution for the issue. If anyone has something that is better please respond 😀

```js
 // import
 import { useRouter } from 'next/router';

 const router = useRouter();

 // in my case I am getting courseId and userId
 const { courseId, userId } = router.query

 // put this line before returning the page JSX.  <-- might be a "hacky" fix
 if (!router.isReady) return null;

 return <>{content}</>
```

**Comment by AnshulShirbhate** [comment]
I have a solution for this problem. So It was basically arising when I was not giving a value in the href attribute of the Link tag of Next.JS. As soon as removed that Link the error disappeared. 100% Working✅✅✅✅

**Comment by salimdriai** [comment]
is there a fix for this ?

**Comment by Douk1NG** [comment]
for router.push, use shallow true, docs: https://nextjs.org/docs/pages/api-reference/functions/use-router
example
`router.push({
            pathname: router.pathname,
            query: {
                id
            }
        }, undefined, { shallow: true });`
